### PR TITLE
Use hconstr to abstract universes

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
category theory failed due to missing overlay for equations vs https://github.com/coq/coq/pull/19313

**Comment by SkySkimmer** [comment]
Are there any other significant users of universe polymorphism?

File: kernel/hConstr.ml [hunk 1]
```diff
@@ -419,3 +419,144 @@ let hcons x =
...
+    | None ->
+      let v = subst_univs_aux tbl subst c in
+      Tbl.add tbl c v;
+      v
+
+and subst_univs_aux tbl subst c =
+  let k = kind c in
+  let k' = map_kind
+      (UVars.subst_sort_level_relevance subst)
+      (UVars.subst_sort_level_instance subst)
```
Comment (SkySkimmer) [inline kernel/hConstr.ml:539]: maybe we need to rehashcons instances

**Comment by SkySkimmer** [comment]
Surprisingly little impact on category-theory considering https://github.com/coq/coq/pull/19297#issuecomment-2207799509 had -5%

**Comment by SkySkimmer** [comment]
"more hcons" previous attempt was incorrect, let's try again

**Comment by ppedrot** [comment]
I think this could still be useful if we get more univ-poly developments, e.g. if unimath decides to slowly remove type-in-type.

**Pull request closed without merging**